src/java.base/share/classes/java/lang/System.java

```diff
@@ -2660,6 +2662,20 @@ public StackWalker newStackWalkerInstance(Set<StackWalker.Option> options,
                                                       Continuation continuation) {
                 return StackWalker.newInstance(options, null, contScope, continuation);
             }
+
+            @Override
+            public int mismatchUTF8(String str, byte[] b, int fromIndex, int toIndex) {
+                byte[] encoded = str.isLatin1() ? str.value() : str.getBytes(UTF_8.INSTANCE);
+                if (false) {
+                    // Arrays.mismatch without the range checks (~5% faster micro getEntryHit)
+                    int aLength = encoded.length;
```

Comment:
Part of the difference you're seeing is due to knowing that you'll be matching the entire length of the first array (`encoded, 0, encoded.length`).

As an experiment I added `Arrays.mismatch(byte[], byte[], int, int)` to mismatch the entire range of the first array argument vs the range of the second and can spot an improvement in affected micros:
```
Benchmark                                     (size)  Mode  Cnt   Score   Error  Units
ArraysMismatch.Char.differentSubrangeMatches      90  avgt   10  12.165 ± 0.074  ns/op # mismatch(a, aFrom, aTo, b, bFrom, bTo)
ArraysMismatch.Char.subrangeMatches               90  avgt   10  10.748 ± 0.006  ns/op # mismatch(a, b, bFrom, bTo)
```
This might be something we can solve in the JITs without having to add new methods to `java.util.Arrays` to deal as efficiently as possible with the case when we're matching against the entirety of one of the arrays.

Reply:
Interesting. Would be nice to solve this in the JIT!

This disabled code got deleted in my last commit, but it seems like you have a good analysis so we can let it go now.

Reply:
Nevermind, I had a flaw in my experiment and seems the first range check in a call like `Arrays.mismatch(encoded, 0, encoded.length, b, off, off+len);` should be eliminated. So perhaps you're seeing the cost of the second range check, which might be unavoidable to be safe (zip meta data could otherwise be doctored to try and perform out of bounds reads via intrinsified code)